src/nvapi/nvapi_d3d_low_latency_device.cpp

```diff
@@ -1,145 +1,65 @@
+    std::mutex NvapiD3dLowLatencyDevice::m_mutex = {};
 
-        if (!lowLatencyDeviceFrameId || !LowLatencyDeviceFrameIdInWindow(lowLatencyDeviceFrameId))
-            return 0;
-
-        return m_applicationIdList[((lowLatencyDeviceFrameId - 1) % applicationIdListSize)];
-    }
-
-    bool LowLatencyFrameIdGenerator::IsRepeatedFrame(uint64_t frameID, uint32_t markerType) {
-        // Should always be within bounds since we drop unsupported marker types in the entrypoints
-        return std::exchange(m_lastFrameId[markerType], frameID) == frameID;
-    }
-
-    bool NvapiD3dLowLatencyDevice::SupportsLowLatency(IUnknown* device) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(device);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
-
-        return d3dLowLatencyDevice->SupportsLowLatency();
-    }
-
-    bool NvapiD3dLowLatencyDevice::SupportsLowLatency(ID3D12CommandQueue* commandQueue) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(commandQueue);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
-
-        return d3dLowLatencyDevice->SupportsLowLatency();
+    void NvapiD3dLowLatencyDevice::Reset() {
+        std::scoped_lock lock{m_mutex};
+        m_lowLatencyDeviceMap.clear();
     }
 
-    bool NvapiD3dLowLatencyDevice::LatencySleep(IUnknown* device) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(device);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
+    static Com<ID3DLowLatencyDevice> GetD3DLowLatencyDevice(IUnknown* device) {
+        Com<ID3DLowLatencyDevice> d3dLowLatencyDevice;
 
-        return SUCCEEDED(d3dLowLatencyDevice->LatencySleep());
-    }
+        if (SUCCEEDED(device->QueryInterface(IID_PPV_ARGS(&d3dLowLatencyDevice))))
+            return d3dLowLatencyDevice;
 
-    bool NvapiD3dLowLatencyDevice::SetLatencySleepMode(IUnknown* device, bool lowLatencyMode, bool lowLatencyBoost, uint32_t minimumIntervalUs) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(device);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
+        if (Com<ID3D11DeviceChild> d3d11DeviceChild; SUCCEEDED(device->QueryInterface(IID_PPV_ARGS(&d3d11DeviceChild)))) {
+            Com<ID3D11Device> d3d11Device;
+            d3d11DeviceChild->GetDevice(&d3d11Device);
+            if (SUCCEEDED(d3d11Device->QueryInterface(IID_PPV_ARGS(&d3dLowLatencyDevice))))
+                return d3dLowLatencyDevice;
+        }
 
-        auto success = SUCCEEDED(d3dLowLatencyDevice->SetLatencySleepMode(lowLatencyMode, lowLatencyBoost, minimumIntervalUs));
-        if (success) {
-            std::scoped_lock lock(m_lowLatencyModeMutex);
-            m_lowLatencyModeMap[device] = lowLatencyMode;
+        if (Com<ID3D12CommandQueue> d3d12CommandQueue; SUCCEEDED(device->QueryInterface(IID_PPV_ARGS(&d3d12CommandQueue)))) {
```

Comment:
Technically the smallest interface we need in this case is `ID3D12DeviceChild`… but I guess it doesn't matter.